### ISSUE TITLE
Remote Shell can not display Unicode characters

### ISSUE THREAD

**Issue by Vigorian** [MaxXor/xRAT#321]
Hallo, found a problem with cmd, on Russian windows(i think whith Arabic and Chinese the same) not correctly display text.

Screanschot: http://rghost.ru/8SwCFwBjt

**Comment by MaxXor** [comment]
Thanks for reporting this issue, I'll fix this asap.

**Comment by MaxXor** [comment]
Can you confirm this is fixed?

**Comment by Vigorian** [comment]
not really, http://rghost.ru/8dcktZPNm.view ,but thanks for try to fix

**Comment by yankejustin** [comment]
Why not make big changes on a different branch and test them extensively before placing right into <code>master</code>? <code>master</code> really should be working almost like production... :(
Anyways, this is an issue with the new .Net Serializer system. It doesn't know how to serialize it correctly...

**Comment by MaxXor** [comment]
@yankejustin I know, sorry. Please back up this and show that it can display it currently with Protobuff.

**Comment by yankejustin** [comment]
Hmmm. There are a few things that need to be fixed. I will be back in a few days because I am actually off to vacation for most of this week.
The issue is also present in ProtoBuf but using .Net Serializer, I can't seem to deserialize the object to see what it does. Deserializing it sends it to a method named <code>ReadIntVar64</code> (or something similar... can't recall the exact name). In this method, it does not correct handle the stream and throws an <code>EndOfStreamException</code>. Not entirely sure how it uses the data it receives on the client's end. It seems like an issue in implementation at the action of sending, because it is correctly placed into a packet.

**Comment by MaxXor** [comment]
@yankejustin I can not confirm this and actually I think this has nothing to do with NetSerializer at all, because NetSerializer has no problem de-/serializing unicode strings.

I will look into this more tomorrow, we'll just need to fix the encoding.

Enjoy your vacation! :relaxed:

**Comment by yankejustin** [comment]
Try to peek <a href="https://github.com/MaxXor/xRAT/blob/0a9f3d0dd722ff5c8af30679d2e1e18e552563f8/Server/Core/ProtoBuf/ProtoReader.cs#L562">where ProtoBuf</a> used to serialize a string.
I haven't looked into how .Net serializer works yet. It deals heavily with using op codes and is a bit difficult to understand for me. I'll look into it further and document it if you wish to keep it.
My second idea is that it has to do with how the data is received. Perhaps more extensible functionality is needed to handle the packet a little bit before it is casted. Perhaps handle the shell packets a bit different around the time when we directly parse the payload to <code>IPacket</code>.
<hr />
Thank you, I will enjoy it. :)

**Comment by yankejustin** [comment]
Another piece of information, I tried to make the packets on both ends send each other data using an array of UTF8-Encoded characters. When the data arrived, they were responsible for taking the encoded bytes and turning it into a string. The result was the exact same for both ProtoBuf and .Net Serializer.

**Comment by d3agle** [comment]
Could this be it? http://stackoverflow.com/a/17177904

**Comment by yankejustin** [comment]
Last piece of information:
<h1>Debugging</h1>
After using a <code>MessageBox</code> to display the values of the payloads, I have noticed that this is (correctly) happening:
1. Server sends payload with the <b>correct</b>encoded value.
2. Client receives the payload with the <b>correct</b> encoded value.
3. The <code>Shell</code> does <b>not</b> show correct encoding. Even after setting <code>StandardOutputEncoding</code> and <code>StandardErrorOutputEncoding</code>, <i>the values of the characters returned in the output streams <b>don't show</b></i>. That is why it is returning a "<i>?</i>".
In short, the server sends the correct value, the client receives the correct value, but the console isn't giving it back as it should.
<h1>Other Issues</h1>
- The console isn't giving anything back to us when I attempt to use WMIC.
    1. Type <code>wmic</code> into the remote shell.
    2. Notice the <i>blank</i> response.
    - This command should change us to something like <code>wmic:root\cli></code>, but it does not.

**Comment by Vigorian** [comment]
hallo, got a solution, the fix from d3agle                     
StandardOutputEncoding = Encoding.GetEncoding(850),
 StandardErrorEncoding = Encoding.GetEncoding(850),
was not fully, the 850 are Multilingual Latin I, i have changed id to 866 (Russian) and it works http://rghost.ru/6v7jQ9wHt.view ,now need to code function where detect the OS language, and send output whith correct encoding.

here is all encoding:
437 (US)
720 (Arabic)
737 (Greek)
775 (Baltic)
850 (Multilingual Latin I)
852 (Latin II)
855 (Cyrillic)
857 (Turkish)
858 (Multilingual Latin I + Euro)
862 (Hebrew)
866 (Russian)

**Comment by d3agle** [comment]
Hi @Vigorian, could you head over to my fork and see if the current commit fixes it for you?

**Comment by Vigorian** [comment]
Hi d3agle, your fix did not fix the problem http://rghost.ru/6N428JNXS.view , but thanks for try to fix it.

**Comment by d3agle** [comment]
@Vigorian, you tried the newest one with `NativeMethods.GetACP()` ?

**Comment by Vigorian** [comment]
yes

**Comment by d3agle** [comment]
Thanks for testing it!  Could you try one more thing and see what the active code page is in the shell by typing 'chcp`?

**Comment by Vigorian** [comment]
ok, http://rghost.ru/7lJ7yMNvF.view

**Comment by d3agle** [comment]
Thank you.  Very strange hmmm... if setting encoding to 866 works, and active code page is 866 but still isn't working.  Odd lol

**Comment by Vigorian** [comment]
no problem, thank you =)

**Comment by d3agle** [comment]
@Vigorian, could you please test my newest commit?  I think this will fix your problem :)

**Comment by Vigorian** [comment]
d3agle, thanks very much, work fine=)

**Comment by d3agle** [comment]
Awesome :) I will send a PR

**Comment by MaxXor** [comment]
Very nice fix, thanks.
